File: src/deploy/yolov5_gesture/python/yolov5_gesture_detector.py
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
#
# SPDX-License-Identifier: Apache-2.0

"""
YOLOv5 Object Detector Implementation
"""

import cv2
import numpy as np
import onnxruntime as ort
import spacemit_ort  # noqa: F401 - register SpaceMITExecutionProvider
from typing import List, Tuple, Any, Dict
from pathlib import Path

from core import BaseModel
from common import (
    xywh2xyxy, scale_coords, box_iou, letterbox
)
from core.python.vision_model_exceptions import (
    ModelNotFoundError,
    ModelLoadError,
    ModelInferenceError,
)


class YOLOv5_GestureDetector(BaseModel):
    """
    YOLOv5 gesture detection model.
    """

    def __init__(self, model_path: str,
                 conf_threshold: float = 0.4,
                 iou_threshold: float = 0.5,
                 num_threads: int = 4,
                 **kwargs):
        """
        Initialize YOLOv5 gesture detector.

        Args:
            model_path: Path to ONNX model file
            conf_threshold: Confidence threshold for filtering detections
            iou_threshold: IoU threshold for NMS
            num_threads: Number of threads for ONNX Runtime
            **kwargs: Additional parameters
        """
        self.conf_threshold = conf_threshold
        self.iou_threshold = iou_threshold
        self.num_threads = num_threads
        super().__init__(model_path, **kwargs)

    def _load_model(self, **kwargs):
        """Load YOLOv5 ONNX model."""
        # Check if model file exists
        model_path_obj = Path(self.model_path)
        if not model_path_obj.exists():
            raise ModelNotFoundError(
                model_path=self.model_path,
                message=f"YOLOv5 模型文件不存在: {self.model_path}"
            )

        # Create session options
        session_options = ort.SessionOptions()
        session_options.intra_op_num_threads = self.num_threads

        # Load model
        try:
            providers = kwargs.get('providers', ['SpaceMITExecutionProvider'])
            self.session = ort.InferenceSession(
                self.model_path,
                sess_options=session_options,
                providers=providers
            )
        except Exception as e:
            raise ModelLoadError(
                model_path=self.model_path,
                reason=str(e),
                message=f"YOLOv5 模型加载失败: {self.model_path} - {e}"
            )

        # Get input/output info
        try:
            self.input_name = self.session.get_inputs()[0].name
            self.output_names = [output.name for output in self.session.get_outputs()]
            # Same as original code: input_shape is (height, width)
            self.input_shape = self.session.get_inputs()[0].shape[2:4]
        except Exception as e:
            raise ModelLoadError(
                model_path=self.model_path,
                reason=f"无法获取模型输入/输出信息: {e}",
                message=f"YOLOv5 模型初始化失败: {self.model_path} - {e}"
            )

    def infer(self, image: np.ndarray) -> List[Dict[str, Any]]:
        # 确保模型已加载（延迟加载支持）
        self._ensure_model_loaded()

        image_shape = image.shape[:2]

        # Preprocess image
        try:
            input_tensor, ratio, pad = self.preprocess(image)
        except Exception as e:
            raise ModelInferenceError(
                model_name="YOLOv5",
                reason=f"图像预处理失败: {e}",
                message=f"YOLOv5 预处理错误: {e}"
            )

        # Run inference
        try:
            outputs = self.session.run(self.output_names, {self.input_name: input_tensor})
        except Exception as e:
            raise ModelInferenceError(
                model_name="YOLOv5",
                reason=f"模型推理失败: {e}",
                message=f"YOLOv5 推理错误: {e}"
            )

        # Postprocess
        try:
            detections = self.postprocess(outputs, self.conf_threshold, self.iou_threshold, image_shape, ratio, pad)
        except Exception as e:
            raise ModelInferenceError(
                model_name="YOLOv5",
                reason=f"结果后处理失败: {e}",
                message=f"YOLOv5 后处理错误: {e}"
            )

        return detections


    # Preprocess image
    def preprocess(self, image: np.ndarray) -> Tuple[np.ndarray, float, Tuple[float, float]]:
        img_resized, r, (dw, dh) = letterbox(image, self.input_shape)
        # BGR to RGB
        img_rgb = cv2.cvtColor(img_resized, cv2.COLOR_BGR2RGB)

        # Normalize to [0, 1]
        img_normalized = img_rgb.astype(np.float32) / 255.0

        # HWC to CHW
        img_chw = np.transpose(img_normalized, (2, 0, 1))

        # Add batch dimension
        img_batch = np.expand_dims(img_chw, axis=0)

        return img_batch, r, (dw, dh)

    def postprocess(
        self,
        outputs: np.ndarray,
        conf_threshold: float,
        iou_threshold: float,
        orig_shape: Tuple[int, int],
        ratio: float,
        pad: Tuple[float, float],
    ) -> List[Dict[str, Any]]:
        # non_max_suppression 返回 List[np.ndarray]，需要取出第一个元素
        detections_list = self.non_max_suppression(outputs[0], conf_threshold, iou_threshold)

        if len(detections_list) == 0 or detections_list[0].size == 0:
            return []

        detections = detections_list[0]  # 取出第一个数组

        # scale_coords 需要 ratio_pad 参数，格式为 ((ratio, ratio), (pad[0], pad[1]))
        ratio_pad = ((ratio, ratio), pad)
        # 缩放坐标到原始图像尺寸
        detections[:, :4] = scale_coords(
            self.input_shape, detections[:, :4], orig_shape, ratio_pad=ratio_pad
        )
        outputs = []
        for i in range(len(detections)):
            d = detections[i]
            outputs.append({
                "bbox": [float(d[0]), float(d[1]), float(d[2]), float(d[3])],
                "class_id": int(d[5]),
                "confidence": float(d[4])
            })
        return outputs

    def non_max_suppression(
        self,
        prediction: np.ndarray,
        conf_thres: float = 0.1,
        iou_thres: float = 0.6,
    ) -> List[np.ndarray]:
        """
        纯 NumPy 实现的非极大值抑制 (NMS)，适用于 ONNX 推理输出。

        参数:
            prediction: 模型的原始输出 (1, N, 85) 或 (1, N, 24) 等。
            conf_thres: 置信度阈值。
            iou_thres: IoU 阈值。
        返回:
            一个包含处理后 NumPy 数组的列表 [detections]，格式为 [x1, y1, x2, y2, conf, cls]。
        """
        if prediction.ndim != 3:
            # 确保输入是 (Batch, N_proposals, N_features)
            return [np.array([])]

        prediction = prediction[0]  # 取出 batch 1

        # 1. 置信度过滤
        # box confidence (第5列, index 4) > conf_thres
        xc = prediction[:, 4] > conf_thres
        x = prediction[xc]

        if x.shape[0] == 0:
            return [np.array([])]

        # 2. 类别置信度 * box置信度（原始 YOLOv5 逻辑）
        x[:, 5:] = x[:, 5:] * x[:, 4:5]

        # 3. XYWH -> XYXY
        box = xywh2xyxy(x[:, :4])

        # 4. 多标签处理 (简化版: 找到每个框的最佳类别)
        # conf: 最佳类别分数 (np.max); j: 最佳类别索引 (np.argmax)
        conf = np.amax(x[:, 5:], axis=1, keepdims=True)
        j = np.argmax(x[:, 5:], axis=1, keepdims=True)

        # 5. 组合结果 [x1, y1, x2, y2, conf, cls]
        x = np.concatenate((box, conf, j.astype(np.float32)), axis=1)

        # 6. 再次过滤 (确保 conf > conf_thres)
        x = x[x[:, 4] > conf_thres]
        n = x.shape[0]

        if n == 0:
            return [np.array([])]

        # 7. 纯 NumPy NMS（手势检测默认使用 class-agnostic NMS，避免不同类别在同一只手上“叠框”）
        boxes = x[:, :4]
        scores = x[:, 4]

        order = scores.argsort()[::-1]
        keep = []
        while order.size > 0:
            i = order[0]
            keep.append(i)
            if order.size == 1:
                break
            iou = box_iou(boxes[i][None, :], boxes[order[1:]])[0]
            inds = np.where(iou <= iou_thres)[0]
            order = order[inds + 1]

        return [x[keep]]

File: src/deploy/yolov5_gesture/cpp/yolov5_gesture_detector.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "yolov5_gesture_detector.h"

#include <cassert>
#include <chrono>
#include <algorithm>
#include <cmath>
#include <iostream>
#include <memory>
#include <stdexcept>
#include <string>
#include <utility>
#include <variant>
#include <vector>

#include "common.h"
#include "vision_model_config.h"
#include "vision_model_factory.h"

namespace vision_deploy {

std::unique_ptr<vision_core::BaseModel> YOLOv5GestureDetector::create(const YAML::Node& config, bool lazy_load) {
    std::string model_path = vision_core::yaml_utils::getString(config, "model_path");
    if (model_path.empty()) {
        throw std::runtime_error("model_path not found in config for YOLOv5GestureDetector");
    }

    YAML::Node default_params = config["default_params"];
    if (!default_params) {
        throw std::runtime_error("default_params not found in config for YOLOv5GestureDetector");
    }

    float conf_threshold = vision_core::yaml_utils::getFloat(default_params, "conf_threshold", 0.4f);
    float iou_threshold = vision_core::yaml_utils::getFloat(default_params, "iou_threshold", 0.5f);
    int num_threads = vision_core::yaml_utils::getInt(default_params, "num_threads", 4);
    std::string provider = vision_core::yaml_utils::getProvider(config);

    return std::make_unique<YOLOv5GestureDetector>(
        model_path, conf_threshold, iou_threshold, num_threads, lazy_load, provider);
}

YOLOv5GestureDetector::YOLOv5GestureDetector(const std::string& model_path,
                                                float conf_threshold,
                                                float iou_threshold,
                                                int num_threads,
                                                bool lazy_load,
                                                const std::string& provider)
    : BaseModel(model_path, lazy_load),
        conf_threshold_(conf_threshold),
        iou_threshold_(iou_threshold),
        num_threads_(num_threads),
        provider_(provider) {
    if (!lazy_load) {
        load_model();
    }
}

void YOLOv5GestureDetector::load_model() {
    if (model_loaded_) {
        return;
    }
    init_session(num_threads_, provider_);
    model_loaded_ = true;
}

cv::Mat YOLOv5GestureDetector::preprocess(const cv::Mat& image) {
    if (image.empty()) {
        throw std::runtime_error("Input image is empty");
    }
    ensure_model_loaded();

    const int inputWidth = static_cast<int>(input_shape_[3]);   // width
    const int inputHeight = static_cast<int>(input_shape_[2]);  // height

    // letterbox() in common converts BGR->RGB and pads to requested shape
    cv::Mat padded = vision_common::letterbox(image, std::make_pair(inputHeight, inputWidth));
    return cv::dnn::blobFromImage(
        padded,
        1.0 / 255.0,
        cv::Size(inputWidth, inputHeight),
        cv::Scalar(0, 0, 0),
        true,   // swapRB (already RGB, but keep consistent with other deploys)
        false,  // crop
        CV_32F);
}

vision_common::DetectionResultList YOLOv5GestureDetector::detect(
    const cv::Mat& image,
    float conf_threshold,
    float iou_threshold) {
    vision_core::ImageInput input;
    input.image = image;
    return detect_input(input, conf_threshold, iou_threshold);
}

vision_common::DetectionResultList
YOLOv5GestureDetector::detect_input(
    const vision_core::ImageInput& input,
    float conf_threshold,
    float iou_threshold) {
    ensure_model_loaded();
    reset_runtime_profile();
    const auto t0 = std::chrono::steady_clock::now();

    const float use_conf = conf_threshold > 0.0f ? conf_threshold : conf_threshold_;
    const float use_iou = iou_threshold > 0.0f ? iou_threshold : iou_threshold_;

    const cv::Size orig_size(
        input.image.cols,
        input.format == vision_core::ImagePixelFormat::kNv12
            ? input.image.rows * 2 / 3
            : input.image.rows);
    const auto t_pre0 = std::chrono::steady_clock::now();
    vision_common::OpenClPreprocessSpec spec;
    spec.output_width = static_cast<int>(input_shape_[3]);
    spec.output_height = static_cast<int>(input_shape_[2]);
    spec.resize_mode =
        vision_common::PreprocessResizeMode::kLetterbox;
    spec.output_rgb = true;
    spec.scale = {
        1.0F / 255.0F,
        1.0F / 255.0F,
        1.0F / 255.0F};
    spec.padding = {114.0F, 114.0F, 114.0F};
    auto prepared = prepare_image(
        input, spec,
        [this](const cv::Mat& bgr) {
            return preprocess(bgr);
        });
    const auto t_pre1 = std::chrono::steady_clock::now();
    set_runtime_preprocess_ms(std::chrono::duration<double, std::milli>(t_pre1 - t_pre0).count());

    const auto t_infer0 = std::chrono::steady_clock::now();
    std::vector<Ort::Value> outputs =
        run_session(prepared.tensor());
    const auto t_infer1 = std::chrono::steady_clock::now();
    set_runtime_model_infer_ms(std::chrono::duration<double, std::milli>(t_infer1 - t_infer0).count());

    const auto t_post0 = std::chrono::steady_clock::now();
    vision_common::DetectionResultList results = postprocess(outputs, orig_size, use_conf, use_iou);
    const auto t_post1 = std::chrono::steady_clock::now();
    set_runtime_postprocess_ms(std::chrono::duration<double, std::milli>(t_post1 - t_post0).count());

    const auto t1 = std::chrono::steady_clock::now();
    set_runtime_total_ms(std::chrono::duration<double, std::milli>(t1 - t0).count());

    return results;
}


std::vector<vision_core::InferIntent> YOLOv5GestureDetector::supported_intents() const {
    return {vision_core::InferIntent::kDetect};
}

vision_core::InferResponse YOLOv5GestureDetector::Run(const vision_core::InferRequest& request) {
    assert(request.intent == vision_core::InferIntent::kDetect);
    const auto* image_input = std::get_if<vision_core::ImageInput>(&request.input);
    if (image_input == nullptr) {
        vision_core::InferResponse response;
        response.ok = false;
        response.error_message = "YOLOv5GestureDetector expects ImageInput";
        return response;
    }

    vision_common::DetectionResultList task_results =
        detect_input(
            *image_input,
            request.params.conf_threshold,
            request.params.iou_threshold);
    vision_core::InferResponse response;
    response.results.reserve(task_results.size());
    for (auto& item : task_results) {
        response.results.emplace_back(std::move(item));
    }
    return response;
}

std::vector<vision_core::ModelCapability> YOLOv5GestureDetector::get_capabilities() const {
    return {vision_core::ModelCapability::kDraw};
}

vision_common::DetectionResultList YOLOv5GestureDetector::postprocess(
    std::vector<Ort::Value>& outputs,
    const cv::Size& orig_size,
    float conf_threshold,
    float iou_threshold) {

    if (outputs.empty()) {
        return {};
    }

    // Model input shape (letterbox output) for clipping before NMS
    const int inputHeight = static_cast<int>(input_shape_[2]);
    const int inputWidth = static_cast<int>(input_shape_[3]);

    // YOLOv5 ONNX typical output: (1, N, 5+nc) with xywh, obj, cls...
    Ort::Value& out0 = outputs[0];
    auto info = out0.GetTensorTypeAndShapeInfo();
    std::vector<int64_t> dims = info.GetShape();
    if (dims.size() < 2) {
        throw std::runtime_error("Unexpected YOLOv5 output shape (dims < 2)");
    }

    // Flatten to [num_boxes, features]
    int64_t num_boxes = 0;
    int64_t features = 0;
    if (dims.size() == 3) {
        // (bs, N, F)
        num_boxes = dims[1];
        features = dims[2];
    } else if (dims.size() == 2) {
        // (N, F)
        num_boxes = dims[0];
        features = dims[1];
    } else {
        // Some exports may provide extra dims; keep it conservative.
        // Try interpret last dim as features, and the product of others (excluding batch) as num_boxes.
        features = dims.back();
        num_boxes = 1;
        for (size_t i = 0; i + 1 < dims.size(); ++i) {
            // skip batch dim if present
            if (i == 0 && dims[0] == 1) continue;
            num_boxes *= dims[i];
        }
    }

    if (features < 6) {
        throw std::runtime_error("Unexpected YOLOv5 output features (< 6)");
    }
    const int64_t num_classes = features - 5;

    const float* data = out0.GetTensorMutableData<float>();

    vision_common::DetectionResultList candidates;
    candidates.reserve(static_cast<size_t>(num_boxes));

    // Iterate proposals
    for (int64_t i = 0; i < num_boxes; ++i) {
        const float* p = data + i * features;

        const float obj = p[4];
        if (obj <= conf_threshold) {
            continue;
        }

        // find best class score (already multiplied by obj in python)
        float best_conf = 0.0f;
        int best_cls = -1;
        for (int64_t c = 0; c < num_classes; ++c) {
            float conf = p[5 + c] * obj;
            if (conf > best_conf) {
                best_conf = conf;
                best_cls = static_cast<int>(c);
            }
        }

        if (best_cls < 0 || best_conf <= conf_threshold) {
            continue;
        }

        // xywh -> xyxy
        float xywh[4] = {p[0], p[1], p[2], p[3]};
        float xyxy[4];
        vision_common::xywh2xyxy(xywh, xyxy);

        // Normalize + clip BEFORE NMS to avoid final clip making boxes collapse into overlaps.
        float x1 = std::min(xyxy[0], xyxy[2]);
        float y1 = std::min(xyxy[1], xyxy[3]);
        float x2 = std::max(xyxy[0], xyxy[2]);
        float y2 = std::max(xyxy[1], xyxy[3]);

        x1 = std::max(0.0f, std::min(x1, static_cast<float>(inputWidth)));
        x2 = std::max(0.0f, std::min(x2, static_cast<float>(inputWidth)));
        y1 = std::max(0.0f, std::min(y1, static_cast<float>(inputHeight)));
        y2 = std::max(0.0f, std::min(y2, static_cast<float>(inputHeight)));

        if (x2 <= x1 || y2 <= y1) {
            continue;
        }

        vision_common::DetectionResult r;
        r.bbox = vision_common::BoundingBox{x1, y1, x2, y2};
        r.score = best_conf;
        r.label = best_cls;
        candidates.push_back(r);
    }

    if (candidates.empty()) {
        return {};
    }

    // NMS (gesture detection uses class-agnostic NMS to avoid overlapped boxes across classes)
    std::vector<cv::Rect2f> boxes;
    std::vector<float> scores;
    boxes.reserve(candidates.size());
    scores.reserve(candidates.size());
    for (const auto& cand : candidates) {
        boxes.emplace_back(cand.bbox.x1, cand.bbox.y1,
            cand.bbox.x2 - cand.bbox.x1, cand.bbox.y2 - cand.bbox.y1);
        scores.push_back(cand.score);
    }
    std::vector<int> keep_indices = vision_common::nms(boxes, scores, iou_threshold);

    vision_common::DetectionResultList results;
    results.reserve(keep_indices.size());
    for (int idx : keep_indices) {
        results.push_back(candidates[idx]);
    }

    // Scale back to original image size (letterbox-aware)
    const cv::Size input_shape(inputWidth, inputHeight);
    for (auto& r : results) {
        float coords[4] = {r.bbox.x1, r.bbox.y1, r.bbox.x2, r.bbox.y2};
        vision_common::scale_coords(input_shape, coords, orig_size);
        r.bbox.x1 = coords[0];
        r.bbox.y1 = coords[1];
        r.bbox.x2 = coords[2];
        r.bbox.y2 = coords[3];
    }

    return results;
}

// Self-registration: YAML class "...YOLOv5_GestureDetector" -> class_name "YOLOv5_GestureDetector"
static vision_core::ModelRegistrar<YOLOv5GestureDetector> registrar("YOLOv5_GestureDetector");

}  // namespace vision_deploy
